From b8569a1eaa70ba4c87f7a5a62019512b3df5e2fc Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Fri, 3 Jan 2025 12:01:07 +0800
Subject: [PATCH 146/206] mtk: hostapd: unset MLE cap bit 3 for Wi-Fi 7 CERT
 4.50

A new configuration option "neg_ttlm_support_mode" is added for the
Wi-Fi 7 CERT 4.50, which requires TID-To-Link Mapping Negotiation
Support subfield in MLD Capabilities And Operations field being 1

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 hostapd/config_file.c   | 2 ++
 src/ap/ap_config.h      | 1 +
 src/ap/ieee802_11_eht.c | 4 ++++
 3 files changed, 7 insertions(+)

diff --git a/hostapd/config_file.c b/hostapd/config_file.c
index 8ac6b2e4a..a01175771 100644
--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -5242,6 +5242,8 @@ static int hostapd_config_fill(struct hostapd_config *conf,
 		conf->eml_disable = atoi(pos);
 	} else if (os_strcmp(buf, "eml_resp") == 0) {
 		conf->eml_resp = atoi(pos);
+	} else if (os_strcmp(buf, "neg_ttlm_support_mode") == 0) {
+		conf->neg_ttlm_support_mode = atoi(pos);
 #ifdef CONFIG_TESTING_OPTIONS
 	} else if (os_strcmp(buf, "eht_oper_puncturing_override") == 0) {
 		if (get_u16(pos, line, &bss->eht_oper_puncturing_override))
diff --git a/src/ap/ap_config.h b/src/ap/ap_config.h
index a2c4d2b31..8a91bcc55 100644
--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -1272,6 +1272,7 @@ struct hostapd_config {
 	u8 eht_bw320_offset;
 	u8 eml_disable;
 	u8 eml_resp;
+	u8 neg_ttlm_support_mode; /* Used for Wi-Fi 7 CERT 4.50 */
 #endif /* CONFIG_IEEE80211BE */
 
 	/* EHT enable/disable config from CHAN_SWITCH */
diff --git a/src/ap/ieee802_11_eht.c b/src/ap/ieee802_11_eht.c
index 21fe9c1d6..4d068edf2 100644
--- a/src/ap/ieee802_11_eht.c
+++ b/src/ap/ieee802_11_eht.c
@@ -599,6 +599,10 @@ static u8 * hostapd_eid_eht_basic_ml_common(struct hostapd_data *hapd,
 	/* TODO: Advertise T2LM based on driver support as well */
 	mld_cap |= EHT_ML_MLD_CAPA_TID_TO_LINK_MAP_ALL_TO_ALL;
 
+	/* For CERT wifi7 4.50 */
+	if (hapd->iconf->neg_ttlm_support_mode == 1)
+		mld_cap &= ~EHT_ML_MLD_CAPA_TID_TO_LINK_MAP_ALL_TO_ONE;
+
 	wpa_printf(MSG_DEBUG, "MLD: MLD Capabilities and Operations=0x%x",
 		   mld_cap);
 	wpabuf_put_le16(buf, mld_cap);
-- 
2.45.2

